cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/spark/data/SparkSchemaCreator.java

```diff
@@ -0,0 +1,191 @@
+import com.ibm.cohort.cql.evaluation.CqlEvaluationRequests;
+import com.ibm.cohort.cql.library.CqlLibraryDescriptor;
+import com.ibm.cohort.cql.library.CqlLibraryProvider;
+import com.ibm.cohort.cql.spark.aggregation.ContextDefinition;
+import com.ibm.cohort.cql.spark.aggregation.ContextDefinitions;
+
+import scala.Tuple2;
+
+
+public class SparkSchemaCreator {
+	private CqlLibraryProvider libraryProvider;
+	private CqlEvaluationRequests requests;
+	private ContextDefinitions contextDefinitions;
+	private SparkOutputColumnNameFactory outputColumnNameFactory;
+
+	public SparkSchemaCreator(CqlLibraryProvider libraryProvider, CqlEvaluationRequests requests, ContextDefinitions contextDefinitions, SparkOutputColumnNameFactory outputColumnNameFactory) {
+		this.libraryProvider = libraryProvider;
+		this.requests = requests;
+		this.contextDefinitions = contextDefinitions;
+		this.outputColumnNameFactory = outputColumnNameFactory;
+	}
+
+	public Map<String, StructType> calculateSchemasForContexts(List<String> contextNames) throws Exception {
+		HashMap<String, StructType> contextResultSchemas = new HashMap<>();
+
+		for (String contextName : contextNames) {
+			StructType schema = calculateSchemaForContext(contextName);
+			if (schema != null) {
+				contextResultSchemas.put(contextName, schema);
+			}
+		}
+		
+		return contextResultSchemas;
+	}
+
+	public StructType calculateSchemaForContext(String contextName) throws Exception {
+		List<CqlEvaluationRequest> filteredRequests = requests.getEvaluationsForContext(contextName);
+
+		StructType resultsSchema = new StructType();
+
+		Set<Tuple2<String, String>> usingInfo = new HashSet<>();
+
+		for (CqlEvaluationRequest filteredRequest : filteredRequests) {
+			CqlLibraryDescriptor descriptor = filteredRequest.getDescriptor();
+			String libraryId = descriptor.getLibraryId();
+
+			for (String expression : filteredRequest.getExpressions()) {
+				Library library = CqlLibraryReader.read(
+						libraryProvider.getLibrary(new CqlLibraryDescriptor().setLibraryId(libraryId).setVersion(descriptor.getVersion())).getContentAsStream()
+				);
+
+				// Track the set of non-system using statements across libraries.
+				// Information is used later to access ModelInfos when searching
+				// for context key column type information.
+				usingInfo.addAll(library.getUsings().getDef().stream()
+										 .filter(x -> !x.getLocalIdentifier().equals("System"))
+										 .map(x -> new Tuple2<>(x.getLocalIdentifier(), x.getVersion()))
+										 .collect(Collectors.toList()));
+
+				List<ExpressionDef> expressionDefs = library.getStatements().getDef().stream()
+						.filter(x -> x.getName().equals(expression))
+						.collect(Collectors.toList());
+
+				if (expressionDefs.isEmpty()) {
+					throw new IllegalArgumentException("Expression " + expression + " is configured in the CQL jobs file, but not found in "
+															   + descriptor.getLibraryId() + "." + descriptor.getVersion());
+				}
```

Comment:
I  had this check previously, but removed it because the cql translator wouldn't let me translate cql with duplicate define names. I can add the check back just in case the translator behavior changes at some point.